scipy/spatial/transform/tests/test_rotation.py

```diff
@@ -455,3 +456,122 @@ def test_from_euler_extrinsic_rotation_313():
     ]))
+
+
+def test_as_euler_asymmetric_axes():
+    np.random.seed(0)
+    n = 10
+    angles = np.empty((n, 3))
+    angles[:, 0] = np.random.uniform(low=-np.pi, high=np.pi, size=(n,))
+    angles[:, 1] = np.random.uniform(low=-np.pi / 2, high=np.pi / 2, size=(n,))
+    angles[:, 2] = np.random.uniform(low=-np.pi, high=np.pi, size=(n,))
+
+    for seq_tuple in permutations('xyz'):
+        # Extrinsic rotations
+        seq = ''.join(seq_tuple)
+        assert_allclose(angles, Rotation.from_euler(seq, angles).as_euler(seq))
+        # Intrinsic rotations
+        seq = seq.upper()
+        assert_allclose(angles, Rotation.from_euler(seq, angles).as_euler(seq))
+
+
+def test_as_euler_symmetric_axes():
+    np.random.seed(0)
+    n = 10
+    angles = np.empty((n, 3))
+    angles[:, 0] = np.random.uniform(low=-np.pi, high=np.pi, size=(n,))
+    angles[:, 1] = np.random.uniform(low=0, high=np.pi, size=(n,))
+    angles[:, 2] = np.random.uniform(low=-np.pi, high=np.pi, size=(n,))
+
+    for axis1 in ['x', 'y', 'z']:
+        for axis2 in ['x', 'y', 'z']:
+            if axis1 == axis2:
+                continue
+            # Extrinsic rotations
+            seq = axis1 + axis2 + axis1
+            assert_allclose(
+                angles, Rotation.from_euler(seq, angles).as_euler(seq))
+            # Intrinsic rotations
+            seq = seq.upper()
+            assert_allclose(
+                angles, Rotation.from_euler(seq, angles).as_euler(seq))
+
+
+def test_as_euler_degenerate_asymmetric_axes():
+    # Since we cannot check for angle equality, we check for dcm equality
+    angles = np.array([
+        [45, 90, 35],
+        [35, -90, 20],
+        [35, 90, 25],
+        [15, 0, 60],
+        [35, 0, 75],
+        [25, -90, 15]
+        ])
+
+    with pytest.warns(UserWarning, match="Gimbal lock"):
+        for seq_tuple in permutations('xyz'):
+            # Extrinsic rotations
+            seq = ''.join(seq_tuple)
+            rotation = Rotation.from_euler(seq, angles, degrees=True)
+            dcm_expected = rotation.as_dcm()
+
+            angle_estimates = rotation.as_euler(seq, degrees=True)
+            dcm_estimated = Rotation.from_euler(
+                seq, angle_estimates, degrees=True
+                ).as_dcm()
+
+            assert_array_almost_equal(dcm_expected, dcm_estimated)
+
+            # Intrinsic rotations
+            seq = seq.upper()
+            rotation = Rotation.from_euler(seq, angles, degrees=True)
+            dcm_expected = rotation.as_dcm()
+
+            angle_estimates = rotation.as_euler(seq, degrees=True)
+            dcm_estimated = Rotation.from_euler(
+                seq, angle_estimates, degrees=True
+                ).as_dcm()
+
+            assert_array_almost_equal(dcm_expected, dcm_estimated)
+
+
+def test_as_euler_degenerate_symmetric_axes():
+    # Since we cannot check for angle equality, we check for dcm equality
+    angles = np.array([
+        [45, 90, 35],
```

Comment:
Here 90 degrees does not create the gimbal lock. However 180 degrees will --- add some of those.